### PR TITLE
Speed up User API

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
Speeds up looking up users in bulk via the users API.

Staging doesn't exhibit the same degree of slowness as does prod, but I did observe an appreciable speed-up.  Loading an example set of users before the change took an average of 2.1 seconds, and with the change, that same user set took an average of 1.3 seconds.

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3954
https://dimagi-dev.atlassian.net/browse/SAAS-15151
This introduces a utility function which takes an iterable of users, fetches user data in chunks, primes the caches, and returns a generator of those users, but where `get_sql_data` doesn't need to do a DB hit.  It's kinda analogous to `prefetch_related`.  This is then applied to the users API to reduce the DB hits to fetch sql data by a factor of 100 (the chunk size).

I thought this was a pretty nice, clean approach to the issue.  The alternative would probably involve a "memoizer" object that gets passed around, and possibly overriding `get_user_data`.  This utility can be used on a list of users in any context and it should just work as a drop-in optimization, without altering the interface in any way.

## Safety Assurance
It's a pretty simple change, and I'm happy with automated testing and a dry run on staging.

### QA Plan
n/a

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by MartinRiese — APPROVED** [review]
neat and clean indeed :)

File: corehq/apps/users/tests/test_user_data.py [hunk 1]
```diff
@@ -129,6 +134,27 @@ def test_migrate_user_no_data(self):
         sql_data = SQLUserData.objects.get(domain=self.domain, user_id=user.user_id)
         self.assertEqual(sql_data.data, {})
 
+    def test_prime_user_data_caches(self):
+        users = [
+            self.make_commcare_user(),
+            self.make_commcare_user(),
+            self.make_commcare_user(),
+            self.make_web_user(),
+            self.make_web_user(),
```
Comment (millerdev) [inline corehq/apps/users/tests/test_user_data.py:143]: I'm not seeing where the Couch data is deleted on test cleanup. Looks like `delete_all_users()` is called in setup, but not during teardown.

File: corehq/apps/users/user_data.py [hunk 2]
```diff
@@ -188,3 +190,23 @@ class SQLUserData(models.Model):
     class Meta:
         unique_together = ("user_id", "domain")
         indexes = [models.Index(fields=['user_id', 'domain'])]
+
+
+def prime_user_data_caches(users, domain):
+    """Lookup and prime caches of user data in chunks"""
```
Comment (millerdev) [inline corehq/apps/users/user_data.py:196]: The name of this function implies that caches are primed when it is called, but nothing will happen unless the generator is iterated. What do you think of adding note about that in the docstring?

File: corehq/apps/users/tests/test_user_data.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/users/tests/test_user_data.py:143]: `make_commcare_user` and `make_web_user` call `addCleanup` on the user.

File: corehq/apps/users/user_data.py (hunk 2 above)
Comment (esoergel) [reply corehq/apps/users/user_data.py:196]: Good point, will do

Comment (esoergel) [reply corehq/apps/users/user_data.py:196]: I wrote it initially as something that would evaluate all up front, but lazy eval seemed more appropriate for an "enrich" type utility.  Let me know if you have a better idea for the name.

**Review by millerdev — APPROVED** [review]
Looks good.

**Pull request merged**